internal/utils/array.go

```diff
@@ -0,0 +1,15 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package utils
+
+// Flatten takes as argument a nested array of elements T.
+// Example: [["a", "b"], ["c", "d"]] -> ["a", "b", "c", "d"]
```

Comment:
**suggestion:** Maybe we should consider to add as dependency some very helpful generic lib like: https://github.com/samber/lo

Reply:
I'd rather not add dependencies for simple things like this.

Reply:
Sure, but above dependencies enable a lot more stuff (similar to Java streams, but embrassing golang, with channel idiom like FanIn/FanOut, or ChannelToSlice/SliceToChannel very helpful in tests or asynchronous orchestration.

I have use it in one of my [personal project](https://pkg.go.dev/github.com/ezian/pipe), and it has reduced the code A LOT.

So I totally agree on your sentences, but if i may suggest to you to double check this library and evaluate it through your knowledge on Juju, it may worth your time ;)